File: src/skill_xray/checks/preproc.py
"""Load-time preprocessing findings derived from parsed Markdown tokens."""

from __future__ import annotations

import hashlib

from ..findings import FINDING_CAP, Finding, sort_findings

_ROOT_KINDS = {"skill_manifest", "agent_identity"}
_LIFTABLE_KINDS = {"instruction", "doc"}
_EVIDENCE_LIMIT = 400


def _loaded_paths(parsed):
    adjacency = {}
    for ref in parsed.refs:
        adjacency.setdefault(ref["from"], []).append(ref["to"])
    loaded = {artifact.rel for artifact in parsed.artifacts if artifact.kind in _ROOT_KINDS}
    queue = list(loaded)
    while queue:
        source = queue.pop()
        for target in adjacency.get(source, []):
            artifact = parsed.by_rel.get(target)
            if target in loaded or artifact is None or artifact.kind not in _LIFTABLE_KINDS:
                continue
            loaded.add(target)
            queue.append(target)
    return loaded


def _bounded_text(value):
    return value[:_EVIDENCE_LIMIT], len(value) > _EVIDENCE_LIMIT


def _inline_finding(artifact, token):
    command = token.code
    if not token.runs or not command.strip():
        return None
    shown, truncated = _bounded_text(command)
    digest = hashlib.sha256(command.encode("utf-8")).hexdigest()
    return Finding(
        vector="SXV-001",
        rule="preproc-inline-bang",
        severity="critical",
        path=artifact.rel,
        line=token.line,
        column=token.column,
        message=(
            "inline preprocessing executes `%s` while loading the skill, before its "
            "instructions are evaluated" % command[:160]
        ),
        evidence={
            "command_text": shown,
            "command_length": len(command),
            "command_sha256": digest,
            "truncated": truncated,
            "column": token.column,
            "fence_state": "outside",
            "selector": "inline-bang:%s" % digest[:12],
        },
    )


def _fenced_finding(artifact, token):
    if not token.code.strip():
        return None
    normalized = token.code.rstrip("\n")
    command_lines = normalized.split("\n")
    executable_lines = sum(bool(line.strip()) for line in command_lines)
    digest = hashlib.sha256(normalized.encode("utf-8")).hexdigest()
    shown, truncated = _bounded_text(normalized)
    return Finding(
        vector="SXV-002",
        rule="preproc-fenced-bang",
        severity="critical",
        path=artifact.rel,
        line=token.line,
        column=token.column,
        message=(
            "bang-tagged fenced preprocessing executes %d command line(s) while "
            "loading the skill" % executable_lines
        ),
        evidence={
            "command_text": shown.split("\n"),
            "command_length": len(normalized),
            "command_sha256": digest,
            "truncated": truncated,
            "column": token.column,
            "fence_info": token.info,
            "block_line_count": executable_lines,
            "selector": "fenced-bang:%s" % digest[:12],
        },
    )


def check(parsed) -> list[Finding]:
    findings = []
    loaded = _loaded_paths(parsed)
    for artifact in parsed.artifacts:
        if artifact.rel not in loaded:
            continue
        for token in artifact.preprocessing:
            if token.kind == "inline":
                if not token.runs or not token.code.strip():
                    continue
                vector = "SXV-001"
            elif token.kind == "fenced":
                if not any(line.strip() for line in token.code.splitlines()):
                    continue
                vector = "SXV-002"
            else:
                continue
            if token.kind == "inline":
                finding = _inline_finding(artifact, token)
            else:
                finding = _fenced_finding(artifact, token)
            if finding is not None:
                findings.append(finding)
        for kind, vector in (("inline", "SXV-001"), ("fenced", "SXV-002")):
            count = artifact.preprocessing_counts[kind]
            if count > FINDING_CAP:
                findings.append(Finding(
                    vector="", rule="findings-capped", severity="low", path=artifact.rel,
                    message="%d more %s findings in %s were suppressed (cap %d per file)"
                            % (count - FINDING_CAP, vector, artifact.rel, FINDING_CAP),
                ))
    return sort_findings(findings)

File: internal/preproc/preproc.go
// Package preproc reports load-time preprocessing findings (SXV-001 inline bang, SXV-002 bang
// fence) from the parsed IR, Python checks/preproc.py.
package preproc

import (
	"crypto/sha256"
	"fmt"
	"strings"
	"unicode/utf8"

	"github.com/traceforce/skill-xray/internal/findings"
	"github.com/traceforce/skill-xray/internal/parse"
	"github.com/traceforce/skill-xray/internal/pytext"
)

const evidenceLimit = 400

var (
	rootKinds     = map[string]bool{"skill_manifest": true, "agent_identity": true}
	liftableKinds = map[string]bool{"instruction": true, "doc": true}
)

func finding(vector, rule, rel string, tok parse.Preproc, message string, evidence map[string]any) findings.Finding {
	return findings.Finding{Vector: vector, Rule: rule, Severity: "critical", Path: rel,
		Line: findings.Int(tok.Line), Column: findings.Int(tok.Column), Message: message, Evidence: evidence}
}

func inlineFinding(rel string, tok parse.Preproc) findings.Finding {
	command := tok.Code
	digest := fmt.Sprintf("%x", sha256.Sum256([]byte(command)))
	return finding("SXV-001", "preproc-inline-bang", rel, tok,
		"inline preprocessing executes `"+pytext.Head(command, 160)+"` while loading the skill, before its instructions are evaluated",
		map[string]any{
			"command_text":   pytext.Head(command, evidenceLimit),
			"command_length": utf8.RuneCountInString(command),
			"command_sha256": digest,
			"truncated":      utf8.RuneCountInString(command) > evidenceLimit,
			"column":         tok.Column,
			"fence_state":    "outside",
			"selector":       "inline-bang:" + digest[:12],
		})
}

func fencedFinding(rel string, tok parse.Preproc) findings.Finding {
	normalized := strings.TrimRight(tok.Code, "\n")
	executable := 0
	for _, line := range strings.Split(normalized, "\n") {
		if pytext.Strip(line) != "" {
			executable++
		}
	}
	digest := fmt.Sprintf("%x", sha256.Sum256([]byte(normalized)))
	shown := strings.Split(pytext.Head(normalized, evidenceLimit), "\n")
	return finding("SXV-002", "preproc-fenced-bang", rel, tok,
		fmt.Sprintf("bang-tagged fenced preprocessing executes %d command line(s) while loading the skill", executable),
		map[string]any{
			"command_text":     shown,
			"command_length":   utf8.RuneCountInString(normalized),
			"command_sha256":   digest,
			"truncated":        utf8.RuneCountInString(normalized) > evidenceLimit,
			"column":           tok.Column,
			"fence_info":       tok.Info,
			"block_line_count": executable,
			"selector":         "fenced-bang:" + digest[:12],
		})
}

// Check is preproc.check: one finding per live, non-blank preprocessing token of every loaded
// artifact, plus a findings-capped note per kind whose parser count exceeded the cap.
func Check(p *parse.Package) []findings.Finding {
	var out []findings.Finding
	// the agent loads manifests and identity files plus every instruction/doc they reach over Refs
	loaded := parse.Reach(p, rootKinds, liftableKinds, true)
	for _, a := range p.Artifacts {
		if !loaded[a.Rel] {
			continue
		}
		for _, tok := range a.Preprocessing {
			switch {
			case pytext.Strip(tok.Code) == "":
			case tok.Kind == "inline" && tok.Runs:
				out = append(out, inlineFinding(a.Rel, tok))
			case tok.Kind == "fenced":
				out = append(out, fencedFinding(a.Rel, tok))
			}
		}
		for _, c := range []struct {
			count  int
			vector string
		}{{a.PreprocessingCounts.Inline, "SXV-001"}, {a.PreprocessingCounts.Fenced, "SXV-002"}} {
			if c.count > findings.Cap {
				out = append(out, findings.Finding{Rule: "findings-capped", Severity: "low", Path: a.Rel,
					Message: fmt.Sprintf("%d more %s findings in %s were suppressed (cap %d per file)",
						c.count-findings.Cap, c.vector, a.Rel, findings.Cap)})
			}
		}
	}
	return findings.Sort(out)
}
